# Khai báo thư viện

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Đọc dữ liệu

In [8]:
data = pd.read_csv('data12012.csv')

In [12]:
data.loc[[591]].T

591
Name                                         iPhone 14 Pro Max
Brand                                                    Apple
Announced                                   2022, September 07
Status                  Available. Released 2022, September 16
Size         6.7 inches, 110.2 cm2 (~88.3% screen-to-body r...
Resolution   1290 x 2796 pixels, 19.5:9 ratio (~460 ppi den...
OS                              iOS 16, upgradable to iOS 16.3
Chipset                                Apple A16 Bionic (4 nm)
Card slot                                                   No
ROM                                                      128.0
RAM                                                        6.0
Num of MC                                                  3.0
Pic MC                                                48,12,12
Video MC     4K@24/25/30/60fps, 1080p@25/30/60/120/240fps, ...
Num of SC                                                  1.0
Pic SC                                                      12
Video SC     4K@24/25/30/60fps, 1080p@25/30/60/120fps, gyro...
Loudspeaker                          Yes, with stereo speakers
3.5mm jack                                                  No
Battery                                                 4323.0
Price           $ 939.95 / € 1,275.99 / £ 1,020.00 / ₹ 127,999

# Mô tả dữ liệu

In [6]:
for c in data.columns:
    print(c, end=', ')

Name, Brand, Announced, Status, Size, Resolution, OS, Chipset, Card slot, ROM, RAM, Num of MC, Pic MC, Video MC, Num of SC, Pic SC, Video SC, Loudspeaker, 3.5mm jack, Battery, Price, 

## Số cột, hàng

In [4]:
n_rows,n_columns = data.shape
print("Số cột:",n_columns, "\nSố hàng:",n_rows)

Số cột: 21 
Số hàng: 12012


## Thông tin các cột

- Name: Tên mẫu sản phẩm
- Brand: Thương hiệu, hãng 
- Announced: Thời gian công bố sản phẩm
- Status: Tình trạng sản phẩm trên thị trường
    + Coming Soon: Hãng công bố chính thức, nhưng chưa có sẵn để bán
    + Cancelled: Hãng chính thức công bố, sau đó chính thức hủy bỏ không ra sản phẩm
    + Available: Hãng vẫn còn sản xuất và bán trên thị trường
    + Discontinued: Hãng ngừng sản xuất, tuy nhiên vẫn có thể tìm kiếm trên thị trường
- Size: Kích cỡ màn hình (inches)
- Resolution: Độ phân giải của màn hình (pixels)
- OS: Hệ điều hành
- Chipset: Chip điện thoại
- Card slot: Có khe cắm thẻ nhớ hay không
- ROM: Dung lượng lưu trữ (GB)
- RAM: Bộ nhớ tạm thời (GB)
- Num of MC: Số camera sau
- Pic MC: Độ phân giải camera sau (MP)
- Video MC: Chất lượng video camera sau (p)
- Num of SC: Số camera trước 
- Pic SC: Độ phân giải camera trước (MP)
- Video SC: Chất lượng video camera trước (p)
- Loudspeaker: Có loa ngoài hay không
- 3.5mm jack: Có jack tai nghe 3.5 không
- Battery: Dung lượng pin điện thoại (mAh)
- Price: Giá của điện thoại (EUR)

## Kiểu dữ liệu

In [5]:
data.dtypes

Name           object
Brand          object
Announced      object
Status         object
Size           object
Resolution     object
OS             object
Chipset        object
Card slot      object
ROM            object
RAM            object
Num of MC      object
Pic MC         object
Video MC       object
Num of SC      object
Pic SC         object
Video SC       object
Loudspeaker    object
3.5mm jack     object
Battery        object
Price          object
dtype: object

- Tất cả đều là ``object``. Nguyên nhân do quá trình cào sử dụng thư viện ``selenium`` và ``.text`` và cắt chuỗi
- Cần chuyển các cột về kiểu số nguyên hoặc số thực như: ``ROM, RAM, Num of MC/SC, Pic MC/SC, Video MC/SC, Battery, Price``
- Các cột kiểu Yes, No cũng có thể chuyển về 1,0 

## Thống kê đơn giản

In [6]:
data.describe().T

count unique                                               top  \
Name         12012  11414                                             Watch   
Brand        12012    119                                           Samsung   
Announced    11998   2133                                          2005, Q1   
Status       12012    982                                      Discontinued   
Size         10803   5054                              5.0 inches, 68.9 cm2   
Resolution   11966   1333  720 x 1280 pixels, 16:9 ratio (~294 ppi density)   
OS            8249   1541                            Android 4.4.2 (KitKat)   
Chipset       6740    876           Qualcomm MSM8916 Snapdragon 410 (28 nm)   
Card slot    12012    159                        microSDHC (dedicated slot)   
ROM          10233     26                                               0.0   
RAM          10233     42                                               0.0   
Num of MC    10686      5                                               1.0   
Pic MC        9658    391                                                 5   
Video MC     10595    526                                               Yes   
Num of SC     7280      4                                               1.0   
Pic SC        5773     92                                                 5   
Video SC      3313     84                                       1080p@30fps   
Loudspeaker  12012     33                                               Yes   
3.5mm jack   11944      9                                               Yes   
Battery      10955    773                                            5000.0   
Price         8041    829                                     About 100 EUR   

             freq  
Name            8  
Brand        1350  
Announced     156  
Status       8405  
Size           38  
Resolution    551  
OS            430  
Chipset       121  
Card slot    2969  
ROM          2996  
RAM          2996  
Num of MC    8326  
Pic MC       1545  
Video MC     2984  
Num of SC    7117  
Pic SC       1339  
Video SC     2220  
Loudspeaker  9095  
3.5mm jack   7615  
Battery       690  
Price         385

- Có rất nhiều cột bị tình trạng Missing Values, đặc biệt là ``Chipset``(44%), ``Pic SC``(52%) và ``Video SC``(72%)
- ``Name``: Có một số sản phẩm tên trùng nhau (*unique* < *count*), vậy chúng có cùng thương hiệu? Xuất hiện giá trị ``Watch`` trong khi chỉ cần dữ liệu về điện thoại
- ``Brand``: Có 119 thương hiệu điện thoại, trong đó Samsung có nhiều sản phẩm nhất
- ``Status``: Giá trị ``Discontinued`` (ngừng sản xuất) xuất hiện 8405 lần, chiếm gần 70% dữ liệu
- ``Resolution``: Bao gồm cả tỉ lệ màn hình (ratio) và mật độ điểm ảnh (ppi). Tỉ lệ màn hình có thể tính từ pixel nên có thể loại bỏ giá trị này. Mật độ điểm ảnh có thể tách thành 1 cột khác nếu đa phần ``Resolution`` đều có giá trị này
- ``ROM`` và ``RAM``: Xuất hiện các giá trị 0 bất thường và đều có cùng tần suất xuất hiện
- Số camera trước và sau đa phần đều là 1
- Các cột chỉ cần 2 giá trị: ``3.5mm jack, Loudspeaker, Card slot, Status`` xuất hiện nhiều giá trị khác nhau

# Các vấn đề về dữ liệu

## Trùng lặp

Kiểm tra trùng lặp bằng kết hợp tên và hãng điện thoại

In [7]:
len(data[data.duplicated(subset=['Name','Brand'])])

0

Không có hãng điện thoại nào có 2 mẫu cùng tên và cũng không có sản phẩm nào giống nhau

Vậy không có bản ghi nào bị trùng lặp

## Giá trị bị thiếu

In [8]:
print(round(data.isnull().sum()/len(data)*100.0,2))

Name            0.00
Brand           0.00
Announced       0.12
Status          0.00
Size           10.06
Resolution      0.38
OS             31.33
Chipset        43.89
Card slot       0.00
ROM            14.81
RAM            14.81
Num of MC      11.04
Pic MC         19.60
Video MC       11.80
Num of SC      39.39
Pic SC         51.94
Video SC       72.42
Loudspeaker     0.00
3.5mm jack      0.57
Battery         8.80
Price          33.06
dtype: float64


- Có 16 cột bị thiếu dữ liệu, trong đó 12 cột bị thiếu trên 10%, 6 cột thiếu trên 20%
- Đa phần dữ liệu đều bị NaN ở dữ liệu liên quan đến camera trước (SC), Chipset, OS và Price
- Vì ``Video SC`` có số lượng NaN quá lớn nên có thể xoá bỏ cột này
- Nếu số lượng camera bằng 0 thì sẽ không có thông tin của Pic và Video.
- Cần tìm nguyên nhân Chipset chiếm gần 44% bị NaN, tương tự với Price, OS

In [205]:
len(data[data.T.isnull().sum()>=6])

3369

Có 3369 bản ghi bị NaN trên 5 cột. Có thể chọn cách xoá các bản ghi này

In [9]:
len(data[data['Num of MC']==0]), len(data[data['Num of SC']==0])

(0, 0)

Không có sản phẩm nào có số camera bằng 0 trong khi trên thực tế có rất nhiều. Vì vậy trường hợp = 0 chính là trường hợp camera bị NaN

### Camera sau (MC)

In [38]:
data['Num of MC'].isnull().sum()

1326

- Num of MC có 1326 giá trị null. Liệu đây là mất dữ liệu hay máy không có camera chính ?

In [39]:
data.Status[data['Num of MC'].isnull()].value_counts()

Discontinued                              1149
Cancelled                                   36
Available. Released 2019, September          6
Available. Released 2020, September 18       4
Available. Released 2019, October            3
                                          ... 
Available. Released 2020, November 14        1
Available. Released 2020, December 29        1
Available. Released 2021, May 20             1
Available. Released 2021, May 29             1
Available. Released 2020, August 05          1
Name: Status, Length: 105, dtype: int64

In [40]:
data.Announced[data['Num of MC'].isnull()].value_counts()

2000                                     82
1999                                     63
2002                                     48
2001                                     48
2004, Q1                                 47
                                         ..
2010, February. Released 2010, March      1
2009, August. Released 2010, February     1
2008, October. Released 2008              1
2008, October. Released 2009, June        1
2007, May. Released 2007                  1
Name: Announced, Length: 374, dtype: int64

Đa phần đều đã dừng sản xuất và đều là các sản phẩm sản xuất từ lâu. Mà trong khoảng thời gian đó đa phần đều là các điện thoại không có Camera. Vậy các giá trị NaN của ``Num of MC`` cần được thay thế bằng số 0

**Vậy kiểm tra xem liệu số Camera chính khác 0 (không NaN) mà ``Pic MC`` và ``Video MC``có NaN**

#### Pic MC

In [31]:
len(data[(data['Pic MC'].isnull()) & (~data['Num of MC'].isnull())])

1028

Có 1028 bản ghi trong đó số Camera chính khác 0 mà độ phân giải Camera chính bị NaN? Liệu có mất mát dữ liệu

In [54]:
data_VGA = data[(data['Pic MC'].isnull()) & (~data['Num of MC'].isnull())]
data_VGA.head(2)

Name    Brand                        Announced        Status  \
84  Pixi 8  alcatel  2014, June. Released 2014, July  Discontinued   
87    2052  alcatel              2014. Released 2014  Discontinued   

                                                 Size  \
84  8.0 inches, 185.6 cm2 (~71.3% screen-to-body r...   
87  1.8 inches, 10.2 cm2 (~21.0% screen-to-body ra...   

                                           Resolution                      OS  \
84  800 x 1280 pixels, 16:10 ratio (~189 ppi density)  Android 4.4.2 (KitKat)   
87                128 x 160 pixels (~114 ppi density)                     NaN   

   Chipset                   Card slot  ROM  ... Num of MC Pic MC Video MC  \
84     NaN  microSDHC (dedicated slot)  0.0  ...       1.0    NaN      Yes   
87     NaN  microSDHC (dedicated slot)  0.0  ...       1.0    NaN      NaN   

   Num of SC Pic SC Video SC Loudspeaker 3.5mm jack Battery          Price  
84       1.0    NaN      NaN         Yes        Yes  3240.0  About 100 EUR  
87       NaN    NaN      NaN         Yes        Yes   750.0   About 50 EUR  

[2 rows x 21 columns]

- Những máy đời cũ 0.3mp sẽ để đặt là VGA (khi thu thập dữ liệu chỉ lấy chữ số nên đã khiến VGA thành NaN)

In [58]:
len(data_VGA[(data_VGA['Pic MC'].isnull()) & (data_VGA['Num of MC'] == '1.0')])

1027

Quả thực có 1027 bản ghi có số camera là 1 nhưng không có thông tin của Pic MC. Vậy sẽ thay 1027 giá trị này thành 0.3MP. Tuy nhiên vẫn còn 1 bản ghi có số camera khác 1

In [56]:
data_VGA[(data_VGA['Pic MC'].isnull()) & (data_VGA['Num of MC'] != '1.0')]

Name    Brand       Announced                         Status  \
994  3x (2018)  alcatel  2018, February  Available. Released 2018, May   

                                                  Size  \
994  5.7 inches, 83.8 cm2 (~76.3% screen-to-body ra...   

                                           Resolution                    OS  \
994  720 x 1440 pixels, 18:9 ratio (~282 ppi density)  Android 7.1 (Nougat)   

                     Chipset                   Card slot   ROM  ... Num of MC  \
994  Mediatek MT6739 (28 nm)  microSDXC (dedicated slot)  16.0  ...       2.0   

    Pic MC Video MC Num of SC Pic SC Video SC Loudspeaker 3.5mm jack Battery  \
994    NaN      NaN       1.0    NaN      NaN         Yes        Yes  3000.0   

             Price  
994  About 130 EUR  

[1 rows x 21 columns]

Điện thoại alcatel 3x(2018) có độ phân giải của Camera sau là 13MP

**Vậy cần thay thế giá trị Pic MC thành 0.3MP và dòng 994 thành 13MP**

#### Video MC

In [47]:
print(len(data[(data['Video MC'].isnull()) & (~data['Num of MC'].isnull())]),'bản ghi bị thiếu thông tin về chất lượng video ở Camera sau')

91 bản ghi bị thiếu thông tin về chất lượng video ở Camera sau


In [60]:
data[(data['Video MC'].isnull()) & (~data['Num of MC'].isnull())].head(4)

Name    Brand                        Announced        Status  \
41          2007  alcatel              2015. Released 2015  Discontinued   
87          2052  alcatel              2014. Released 2014  Discontinued   
215          Net  alcatel  2010, June. Released 2010, July  Discontinued   
232  OT-802 Wave  alcatel    2009. Released 2009, December  Discontinued   

                                                  Size  \
41   2.4 inches, 17.8 cm2 (~30.0% screen-to-body ra...   
87   1.8 inches, 10.2 cm2 (~21.0% screen-to-body ra...   
215  2.4 inches, 17.8 cm2 (~27.6% screen-to-body ra...   
232  2.4 inches, 17.8 cm2 (~27.6% screen-to-body ra...   

                                         Resolution   OS Chipset  \
41   240 x 320 pixels, 4:3 ratio (~167 ppi density)  NaN     NaN   
87              128 x 160 pixels (~114 ppi density)  NaN     NaN   
215  320 x 240 pixels, 4:3 ratio (~167 ppi density)  NaN     NaN   
232  320 x 240 pixels, 4:3 ratio (~167 ppi density)  NaN     NaN   

                                     Card slot  ROM  ... Num of MC Pic MC  \
41                                          No  0.0  ...       1.0     15   
87                  microSDHC (dedicated slot)  0.0  ...       1.0    NaN   
215  microSDHC (dedicated slot), 4 GB included  0.0  ...       1.0      2   
232  microSDHC (dedicated slot), 4 GB included  0.0  ...       1.0      2   

    Video MC Num of SC Pic SC Video SC Loudspeaker 3.5mm jack Battery  \
41       NaN       NaN    NaN      NaN         Yes         No   750.0   
87       NaN       NaN    NaN      NaN         Yes        Yes   750.0   
215      NaN       NaN    NaN      NaN         Yes         No   850.0   
232      NaN       NaN    NaN      NaN         Yes         No   850.0   

            Price  
41            NaN  
87   About 50 EUR  
215           NaN  
232           NaN  

[4 rows x 21 columns]

Dựa theo thông tin từ nhiều nguồn trên internet, các máy trên đều không chung 1 chất lượng Video, vì vậy để xử lý nhanh chóng, có thể chọn giải pháp xoá các bản ghi này

### Camera trước (SC)

In [64]:
data['Num of SC'].isnull().sum()

4732

``Num of SC`` có 4732 giá trị NaN. Tương tự như Camera sau, liệu đây là mất dữ liệu hay thật sự máy không có camera selfie ?

In [65]:
data['Pic SC'].isnull().sum()

6239

Ta có thể giải thích được 4732 giá trị null ở cột ``Pic SC`` vì nó cũng có giá trị null ở cột ``Num of SC``. Cần thay thế các giá trị này bằng 0

In [75]:
data_PicSC = data[data['Pic SC'].isnull() & (~data['Num of SC'].isnull())]
data_PicSC.head(2)

Name    Brand                         Announced  \
0         1T 7  alcatel                    2018, February   
28  Pixi 4 (7)  alcatel  2016, January. Released 2016, Q3   

                            Status  \
0   Available. Released 2018, June   
28                    Discontinued   

                                                 Size  \
0   7.0 inches, 137.9 cm2 (~65.6% screen-to-body r...   
28  7.0 inches, 137.9 cm2 (~67.7% screen-to-body r...   

                                          Resolution  \
0   600 x 1024 pixels, 16:9 ratio (~170 ppi density)   
28  600 x 1024 pixels, 16:9 ratio (~170 ppi density)   

                           OS                     Chipset  \
0          Android 8.1 (Oreo)  Mediatek MT8321A/D (28 nm)   
28  Android 6.0 (Marshmallow)     Mediatek MT8321 (28 nm)   

                     Card slot  ROM  ... Num of MC Pic MC    Video MC  \
0   microSDXC (dedicated slot)  8.0  ...       1.0      2  720p@30fps   
28  microSDHC (dedicated slot)  8.0  ...       1.0      2  720p@30fps   

   Num of SC Pic SC Video SC Loudspeaker 3.5mm jack Battery         Price  
0        1.0    NaN      NaN         Yes        Yes  2580.0  About 70 EUR  
28       1.0    NaN      NaN         Yes        Yes  2580.0  About 80 EUR  

[2 rows x 21 columns]

In [76]:
len(data_PicSC)

1507

Có 1507 giá trị mà số Camera sau khác 0 nhưng Pic SC NaN. Những máy đời cũ 480p sẽ để đặt là VGA (khi thu thập dữ liệu đã mặc định đổi VGA thành null)

In [77]:
data_PicSC[(data_PicSC['Pic SC'].isnull()) & (data_PicSC['Num of SC'] != '1.0')]

Name    Brand          Announced  \
752       Zenfone 8 Flip     Asus       2021, May 12   
757        Zenfone 7 Pro     Asus    2020, August 26   
758            Zenfone 7     Asus    2020, August 26   
762    Zenfone 6 ZS630KL     Asus       2019, May 16   
7155          Galaxy A80  Samsung     2019, April 10   
10824   NEX Dual Display     vivo     2018, December   
11341       Mi Mix Alpha   Xiaomi    2019, September   
11750          nubia Z20      ZTE       2019, August   
11762            nubia X      ZTE      2018, October   
12011         nubia X 5G      ZTE  Not announced yet   

                                       Status  \
752          Available. Released 2021, May 13   
757    Available. Released 2020, September 01   
758    Available. Released 2020, September 01   
762             Available. Released 2019, May   
7155            Available. Released 2019, May   
10824      Available. Released 2018, December   
11341                               Cancelled   
11750        Available. Released 2019, August   
11762      Available. Released 2018, November   
12011                               Cancelled   

                                                    Size  \
752    6.67 inches, 107.4 cm2 (~84.2% screen-to-body ...   
757    6.67 inches, 107.4 cm2 (~84.2% screen-to-body ...   
758    6.67 inches, 107.4 cm2 (~84.2% screen-to-body ...   
762    6.4 inches, 100.5 cm2 (~83.8% screen-to-body r...   
7155   6.7 inches, 108.4 cm2 (~85.8% screen-to-body r...   
10824  6.39 inches, 100.2 cm2 (~84.7% screen-to-body ...   
11341  7.92 inches, 201.8 cm2 (~180.8% screen-to-body...   
11750  6.42 inches, 101.2 cm2 (~84.7% screen-to-body ...   
11762  6.26 inches, 97.8 cm2 (~86.6% screen-to-body r...   
12011                              6.26 inches, 97.8 cm2   

                                              Resolution  \
752    1080 x 2400 pixels, 20:9 ratio (~395 ppi density)   
757    1080 x 2400 pixels, 20:9 ratio (~395 ppi density)   
758    1080 x 2400 pixels, 20:9 ratio (~395 ppi density)   
762    1080 x 2340 pixels, 19.5:9 ratio (~403 ppi den...   
7155   1080 x 2400 pixels, 20:9 ratio (~393 ppi density)   
10824  1080 x 2340 pixels, 19.5:9 ratio (~403 ppi den...   
11341              2088 x 2250 pixels (~388 ppi density)   
11750  1080 x 2340 pixels, 19.5:9 ratio (~401 ppi den...   
11762  1080 x 2280 pixels, 19:9 ratio (~403 ppi density)   
12011  1080 x 2280 pixels, 19:9 ratio (~403 ppi density)   

                                                      OS  \
752                 Android 11, upgradable to Android 13   
757          Android 10, upgradable to Android 12, ZenUI   
758          Android 10, upgradable to Android 12, ZenUI   
762    Android 9.0 (Pie), upgradable to Android 11, Z...   
7155   Android 9.0 (Pie), upgradable to Android 10, O...   
10824                    Android 9.0 (Pie), Funtouch 4.5   
11341                                Android 10, MIUI 11   
11750                        Android 9.0 (Pie), Nubia UI   
11762                     Android 8.1 (Oreo), Nubia UI 6   
12011                                  Android 9.0 (Pie)   

                                          Chipset                   Card slot  \
752      Qualcomm SM8350 Snapdragon 888 5G (5 nm)  microSDXC (dedicated slot)   
757    Qualcomm SM8250 Snapdragon 865 5G+ (7 nm+)  microSDXC (dedicated slot)   
758     Qualcomm SM8250 Snapdragon 865 5G (7 nm+)  microSDXC (dedicated slot)   
762         Qualcomm SM8150 Snapdragon 855 (7 nm)  microSDXC (dedicated slot)   
7155        Qualcomm SDM730 Snapdragon 730 (8 nm)                          No   
10824      Qualcomm SDM845 Snapdragon 845 (10 nm)                          No   
11341      Qualcomm SM8150 Snapdragon 855+ (7 nm)                          No   
11750      Qualcomm SM8150 Snapdragon 855+ (7 nm)                          No   
11762      Qualcomm SDM845 Snapdragon 845 (10 nm)                          No   
12011       Qualcomm SM8150 Snapdragon 855 (7 nm)                     

Có 10 máy có số lượng camera khác 1, đều là sản phẩm mới gần đây, thậm chí có cả sản phẩm chưa công bố. Theo tìm hiểu thì các mẫu điện thoại này đều là các mẫu có 2 màn hình (trước,sau) hoặc màn hình tràn viền 180%(Mi Mix Alpha). Vì vậy camera trước và sau là một. Bỏi vậy sẽ thay thế các giá trị NaN ở 10 bản ghi này bằng giá trị ở cột Camera sau.

In [79]:
lst_MC_SC = data_PicSC[(data_PicSC['Pic SC'].isnull()) & (data_PicSC['Num of SC'] != '1.0')].index

**Cột ``Video SC`` do thiếu quá nhiều nên chọn phương pháp xoá cột. Vì vậy không cần kiểm tra cột này**

### Announced

In [86]:
len(data[data.Announced.isnull()])

14

In [96]:
data.Status[data.Announced.isnull()].value_counts()

Discontinued    8
Cancelled       6
Name: Status, dtype: int64

- Announced có 14 giá trị NaN
- Tất cả đều trong tình trạng ngừng sản xuất. Vì vậy có thể loại bỏ 14 giá trị này

### ROM, RAM
Số lượng bị NaN đều bằng nhau, có thể trên cùng 1 bản ghi

In [161]:
data[['RAM','ROM']].isnull().sum()

RAM    1779
ROM    1779
dtype: int64

In [162]:
data_RR = data[data.RAM.isnull() & data.ROM.isnull()]

In [163]:
len(data_RR)

1779

In [165]:
data_RR.Status.value_counts()

Discontinued                              1664
Cancelled                                   71
Available. Released 2021, April              4
Available. Released 2020, October 27         3
Available. Released 2018, December           2
Available. Released 2021, September 27       2
Available. Released 2021, March 31           1
Available. Released 2017, November           1
Available. Released 2021, June 13            1
Available. Released 2022, December 09        1
Available. Released 2022, November 10        1
Available. Released 2022, November 12        1
Available. Released 2021, March 11           1
Available. Released 2022, December 11        1
Available. Released 2017, September          1
Available. Released 2020, December 01        1
Available. Released 2020, November 01        1
Available. Released 2020, October 13         1
Available. Released 2020, October 06         1
Available. Released 2020, January            1
Available. Released 2021, June 25            1
Available. Re

Đa phần đều là các sản phẩm đã dừng sản xuất 

In [168]:
data_RR[(data_RR.Status != 'Discontinued') & (data_RR.Status !='Cancelled')]

Name      Brand           Announced  \
376          Allwatch Hybrid T    Allview      2018, December   
377          Allwatch Hybrid S    Allview      2018, December   
380                 Allwatch V    Allview          2018, June   
5719                   Watch 3     Realme       2022, July 26   
5721       TechLife Watch R100     Realme       2022, June 23   
5740       TechLife Watch S100     Realme      2022, March 10   
5754                  Watch T1     Realme    2021, October 19   
5774               Watch 2 Pro     Realme        2021, May 20   
5775                   Watch 2     Realme      2021, April 30   
5794               Watch S Pro     Realme   2020, December 23   
5797                   Watch S     Realme   2020, November 03   
5825                     Watch     Realme        2020, May 25   
6091                       D11     Doogee       2022, July 25   
6093                       CS3     Doogee       2022, July 09   
6101                   CR1 Pro     Doogee      2021, November   
6103                   DG Ares     Doogee  2021, September 27   
6110                       CS2     Doogee           2021, May   
6112                       CR1     Doogee      2021, March 29   
6115                       CS1     Doogee      2021, February   
6149                       E28  Energizer         2021, April   
6150                       E24  Energizer         2021, April   
6151                        E3  Energizer         2021, April   
6152                        E2  Energizer         2021, April   
6186         Energy 100 (2017)  Energizer          2017, June   
6188                Energy E10  Energizer     2017, September   
6189                Energy E20  Energizer      2017, November   
6340              Smartwatch 7     Gionee       2021, June 09   
6998                Watch Buds     Huawei   2022, December 09   
9001             Watch GT 3 SE     Huawei   2022, November 10   
9003            Watch GT Cyber     Huawei   2022, November 02   
9031            Watch Fit mini     Huawei  2021, September 27   
10349                Watch Pro    Ulefone    2020, October 27   
10350                Watch GPS    Ulefone    2020, October 27   
10351                    Watch    Ulefone    2020, October 27   
11147                 Watch S2     Xiaomi   2022, December 11   
11246  Mi Watch Revolve Active     Xiaomi       2021, June 22   
11283              Redmi Watch     Xiaomi   2020, November 26   
11286    Mi Watch Color Sports     Xiaomi    2020, October 20   
11291                 Mi Watch     Xiaomi  2020, September 30   
11292         Mi Watch Revolve     Xiaomi  2020, September 29   
11337              Watch Color     Xiaomi       2020, January   
11714                 Watch GT        ZTE      2021, March 31   
11715          Red Magic Watch        ZTE      2021, March 04   
11721               Watch Live        ZTE   2020, November 27   

                                       Status  \
376        Available. Released 2018, December   
377        Available. Released 2018, December   
380            Available. Released 2018, June   
5719      Available. Released 2022, August 02   
5721        Available. Released 2022, July 28   
5740       Available. Released 2022, March 14   
5754    Available. Released 2021, November 01   
5774         Available. Released 2021, May 29   
5775         Available. Released 2021, May 20   
5794    Available. Released 2020, December 29   
5797    Available. Released 2020, November 14   
5825        Available. Released 2020, June 05   
6091        Available. Released 2022, July 25   
6093        Available. Released 2022, July 09   
6101       Available. Released 2021, November   
6103   Available. Released 2021, September 27   
6110            Available. Released 2021, May   
6112       Available. Released 2021, March 29   
6115       Available. Released 2021, February   
6149          Available. Released 2021, April   
6150          Available. Released 2021, April   
6151          Availabl

Các sản phẩm mà vẫn còn sản xuất thì lại là đồng hồ thông minh

In [179]:
len(data_RR[data_RR.T.isnull().sum() >= 8])

1493

Có 1493 bản ghi bị NaN trên 8 cột

**Vậy cần xoá các bản ghi bị thiếu thông tin Ram, Rom**

### Chipset

In [225]:
data_Chip = data[data.Chipset.isnull()]
len(data_Chip)

5272

44% dữ liệu về chip bị thiếu. Nguyên nhân có thể tiếp tục là do sản phẩm từ lâu, dừng sản xuất

In [227]:
data_Chip.Status.value_counts()

Discontinued                             4851
Cancelled                                 168
Available. Released 2017                   10
Available. Released 2019, July              9
Available. Released 2019, September         6
                                         ... 
Available. Released 2020, June 05           1
Available. Released 2020, November 14       1
Available. Released 2020, December 29       1
Available. Released 2021, May 20            1
Available. Released 2016, July              1
Name: Status, Length: 154, dtype: int64

### Các sản phẩm ngừng sản xuất
Kiểm tra số lượng bị thiếu với các sản phảm ngừng sản xuất

In [243]:
data_DisC = data[(data.Status == 'Discontinued') | (data.Status == 'Cancelled')]
len(data_DisC)

8689

Có 72.3% dữ liệu ngừng sản xuất

In [250]:
data_DisC = data[(data.Status == 'Discontinued') | (data.Status == 'Cancelled')]
data_DisC.isnull().sum()/data.isnull().sum()*100

Name                  NaN
Brand                 NaN
Announced      100.000000
Status                NaN
Size           100.000000
Resolution     100.000000
OS              98.671273
Chipset         95.201062
Card slot             NaN
ROM             97.526700
RAM             97.526700
Num of MC       89.366516
Pic MC          92.990654
Video MC        88.779111
Num of SC       95.984784
Pic SC          96.553935
Video SC        91.297850
Loudspeaker           NaN
3.5mm jack      98.529412
Battery         96.877956
Price           92.848149
dtype: float64

- Có thể thấy số giá trị NaN của tập data_DisC/ số giá trị NaN tập gốc đều trên 88%

**KẾT LUẬN**

- Dữ liệu bị mất mát nhiều do đa phần đều là các sản phẩm ngừng sản xuất, đều là sản phẩm từ lâu
- Việc tìm dữ liệu thay thế của hơn 8000 sản phẩm không khả thi do dữ liệu cần từ nhiều nguồn mà mỗi nguồn cần có phương pháp cào khác nhau.
- Vần xử lý, loại bỏ các trường hợp bị NaN để thuận tiện cho các quá trình sau

# Làm sạch lần 1
Loại bỏ giá trị NaN và thay thế giá trị

## Thay thế giá trị trước khi drop

Không thay thế giá trị NaN ở cột ``Num of MC/SC`` thành 0 bởi cũng có thể có nguyên nhân web chưa có dữ liệu. 

Vì vậy chỉ thay thế bản ghi có giá trị ``Num of MC/SC`` là 1 và giá trị NaN ở ``Pic MC/SC``

In [270]:
dataNew = data.copy()

### Camera sau (MC)

In [271]:
VGA = '0.3'

In [268]:
idx_VGA = data_VGA[(data_VGA['Num of MC'] == '1.0')].index

In [272]:
for i in idx_VGA:
    dataNew.at[i,'Pic MC'] = VGA

In [278]:
dataNew.loc[[994]]

Name    Brand       Announced                         Status  \
994  3x (2018)  alcatel  2018, February  Available. Released 2018, May   

                                                  Size  \
994  5.7 inches, 83.8 cm2 (~76.3% screen-to-body ra...   

                                           Resolution                    OS  \
994  720 x 1440 pixels, 18:9 ratio (~282 ppi density)  Android 7.1 (Nougat)   

                     Chipset                   Card slot   ROM  ... Num of MC  \
994  Mediatek MT6739 (28 nm)  microSDXC (dedicated slot)  16.0  ...       2.0   

    Pic MC     Video MC Num of SC Pic SC Video SC Loudspeaker 3.5mm jack  \
994   13,5  1080p@30fps       1.0      5      NaN         Yes        Yes   

    Battery          Price  
994  3000.0  About 130 EUR  

[1 rows x 21 columns]

Tại index 994, cần thay thành 13, 5 ở Pic MC, 1080p ở Video MC, 5MP ở Pic SC

In [277]:
dataNew.at[994,'Pic MC'] = '13,5'
dataNew.at[994,'Video MC'] = '1080p@30fps'
dataNew.at[994,'Pic SC'] = '5'

### Camera trước (SC)
Một số máy có 2 màn hình sẽ chỉ có 1 cụm camera đóng vai trò làm cả Camera trước và sau. 

Vì vậy thay giá trị ``Pic SC`` giống với ``Pic MC``, ``Video SC`` giống với ``Video MC``

In [265]:
lst_MC_SC

Int64Index([752, 757, 758, 762, 7155, 10824, 11341, 11750, 11762, 12011], dtype='int64')

In [280]:
for i in lst_MC_SC:
    dataNew.at[i,'Pic SC'] = dataNew.at[i,'Pic MC']
    dataNew.at[i,'Video SC'] = dataNew.at[i,'Video MC'] 

Các máy có chất lượng ảnh VGA (0.3MP)

In [263]:
idx_VGA_SC = data_PicSC[data_PicSC['Num of SC'] == '1.0'].index
idx_VGA_SC

Int64Index([    0,    28,    32,    50,    57,    58,    59,    60,    63,
               64,
            ...
            11997, 12000, 12001, 12002, 12003, 12004, 12005, 12006, 12007,
            12008],
           dtype='int64', length=1497)

In [282]:
for i in idx_VGA_SC:
    dataNew.at[i,'Pic SC'] = VGA

## Drop cột Video SC và giá trị NaN

In [284]:
df = dataNew.drop(columns=['Video SC']).dropna()

In [286]:
df.shape

(4884, 20)

Vậy tập dữ liệu mới đã loại bỏ 1 cột và còn 4884 dữ liệu không bị NaN

# Xuất file csv trước khi tiếp tục làm sạch
Giữ lại index cũ

In [287]:
df.to_csv('data4884.csv')

<font size = '5'> **Dữ liệu mới đã không bị trùng lặp và bị NaN. Tiếp tục kiểm tra các vấn đề còn lại** <font>